First steps towards template type-checking in ngtsc

alxhub opened the pull request:
This extra-large PR implements major components of template type-checking in ngtsc. Broadly, it includes:

* Refactors to prepare for the work
* A template binding API to extract semantic information about templates
* A type-checking code generator
* Integration of the above into ngtsc via a second `ts.Program` created for the purposes of type checking.
* Tests for the above

The work is split into multiple commits with a narrow enough scope to be reviewed separately. Relevant tests were verified to pass at each commit.

alxhub committed:
refactor(ivy): extract directive matching code into a utility function

Upcoming implementation work for template type-checking will need to reuse the
code which matches directives inside a template, so this refactor commit moves
the code to a shared location in preparation.

This commit pulls the code needed to match directives against a template node
out of the TemplateDefinitionBuilder into a utility function, in preparation
for template type-checking and other TemplateDefinitionBuilder refactoring.

alxhub committed:
feat(ivy): add generics to the SelectorMatcher API

This commit adds a generic type parameter to the SelectorMatcher
class and its associated response types. This makes the API for
matching selectors and obtaining information about the matched
directives significantly more ergonomic and type-safe.

alxhub committed:
fix(ivy): remove metadata from *Def and introduce *DefWithMeta types

Previously in Ivy, metadata for directives/components/modules/etc was
carried in .d.ts files inside type information encoded on the
DirectiveDef, ComponentDef, NgModuleDef, etc types of Ivy definition
fields. This works well, but has the side effect of complicating Ivy's
runtime code as these extra generic type parameters had to be specified
as <any> throughout the codebase. *DefInternal types were introduced
previously to mitigate this issue, but that's the wrong way to solve
the problem.

This commit returns *Def types to their original form, with no metadata
attached. Instead, new *DefWithMeta types are introduced that alias the
plain definition types and add extra generic parameters. This way the
only code that needs to deal with the extra metadata parameters is the
compiler code that reads and writes them - the existence of this metadata
is transparent to the runtime, as it should be.

alxhub committed:
refactor(ivy): move the expr/stmt translator to a separate target

Template type-checking will make use of expression and statement
translation as well as the ImportManager, so this code needs to
live in a separate build target which can be depended on by both
the main ngtsc transform as well as the template type-checking
mechanism. This refactor introduces a separate build target
for that code.

alxhub committed:
feat(ivy): augment selector scopes to extract additional metadata

Before type checking can be turned on in ngtsc, appropriate metadata for
each component and directive must be determined. This commit adds tracking
of the extra metadata in *DefWithMeta types to the selector scope handling,
allowing for later extraction for type-checking purposes.

alxhub committed:
feat(ivy): turn on template type-checking via fullTemplateTypeCheck

This commit enables generation and checking of a type checking ts.Program
whenever the fullTemplateTypeCheck flag is enabled in tsconfig.json. It
puts together all the pieces built previously and causes diagnostics to be
emitted whenever type errors are discovered in a template.

Todos:

* map errors back to template HTML
* expand set of type errors covered in generated type-check blocks

alxhub committed:
feat(ivy): introduce a new compiler API for operating on templates

This commit introduces the "t2" API, which processes parsed template ASTs
and performs a number of functions such as binding (the process of
semantically interpreting cross-references within the template) and
directive matching. The API is modeled on TypeScript's TypeChecker API,
with oracle methods that give access to collected metadata.

This work is a prerequisite for the upcoming template type-checking
functionality, and will also become the basis for a refactored
TemplateDefinitionBuilder.

alxhub committed:
feat(ivy): introduce the type check block compiler

This commit introduces the main functionality of the type-check compiler:
generation of type check blocks. Type check blocks are blocks of TypeScript
code which can be inlined into source files, and when processed by the
TypeChecker will give information about any typing errors in template
expressions.

alxhub committed:
feat(ivy): introduce type-checking context API

This commit introduces the template type-checking context API, which manages
inlining of type constructors and type-check blocks into ts.SourceFiles.
This API will be used by ngtsc to generate a type-checking ts.Program.

An TypeCheckProgramHost is provided which can wrap a normal ts.CompilerHost
and intercept getSourceFile() calls. This can be used to provide source
files with type check blocks to a ts.Program for type-checking.

alxhub committed:
test(ivy): refactor ngtsc tests to use an NgtscTestEnvironment helper

This commit gets ready for the introduction of ngtsc template
type-checking tests by refactoring test environment setup into a
custom helper. This helper will simplify the authoring of future
ngtsc tests.

Ngtsc tests previously returned a numeric error code (a la ngtsc's CLI
interface) if any TypeScript errors occurred. The helper has the
ability to run ngtsc and return the actual array of ts.Diagnostics, which
greatly increases the ability to write clean tests.

alxhub committed:
test(ivy): implement tests for template type-checking

This commit builds on the NgtscTestEnvironment helper work before and
introduces template_typecheck_spec.ts, which contains compiler tests
for template type-checking.

alxhub committed:
feat(ivy): introduce typecheck package and a type constructor generator

This commit introduces //packages/compiler-cli/src/ngtsc/typecheck as a
container for template type-checking code, and implements an initial API:
type constructor generation.

Type constructors are static methods on component/directive types with
no runtime implementation. The methods are used during compilation to
enable inference of a component or directive's generic type parameters
from the types of expressions bound to any of their @Inputs. A type
constructor looks like:

class Directive<T> {
  someInput: T;
  static ngTypeCtor<T>(init: Partial<Pick<Directive<T>, 'someInput'>>): Directive<T>;
}

It can be used to infer a type for T based on the input:

const _dir = Directive.ngTypeCtor({someInput: 'string'}); // Directive<T>

mary-poppins commented:
You can preview d1a608d at https://pr26203-d1a608d.ngbuilds.io/.

mhevery commented on packages/compiler/src/render3/view/t2_api.ts:
```diff
@@ -0,0 +1,142 @@
+   * A primary directive is one that becomes the default value for a reference to the element on
+   * which it's created. @Components are primary directives, @Directives are not.
+   */
+  isPrimary: boolean;
```
Why not `isComponent`?

alxhub replied:
Done.

mhevery commented on packages/compiler-cli/test/ngtsc/template_typecheck_spec.ts:
```diff
@@ -0,0 +1,172 @@
+    class Module {}
+    `);
+
+    env.driveMain();
```
There does not seem to be an assertion. If you checking that it does not throw, say so.

alxhub replied:
`driveMain` asserts that the compiler ran successfully.

trotyl commented on packages/compiler-cli/test/ngtsc/template_typecheck_spec.ts:
```diff
@@ -0,0 +1,172 @@
+
+export declare class NgIf {
+  ngIf: any;
+  static ngTemplateGuard_ngIf<E>(dir: NgIf, expr: E): expr is NonNullable<E>
```
The semantic for `NonNullable` doesn't seem right for `NgIf`, should be some kind of `NonFalsy` instead, like:

```typescript
type Falsy = false | null | undefined | 0 | ''

class NgIf {
  static ngTemplateGuard_ngIf<E>(dir: NgIf, expr: E | Falsy): expr is E
  // or
  static ngTemplateGuard_ngIf<E>(dir: NgIf, expr: E): expr is Exclude<E, Falsy>
}
```

alxhub replied:
I agree in principle. For the real `NgIf` implementation, we may do something like this. However, the important case is to capture the narrowing to `NonNullable`, because that supports `strictNullChecks` in the generated type-checking code.

Narrowing other false-y values is useful only in very specific situations (calling methods which only accept truthy types). I expect this to be almost non-existent in practice.

BTW, only your latter version of  `ngTemplateGuard_ngIf` will actually work. The former works well for inferred values of `E`, but fails if passed an `expr` with an explicitly nullable type.

trotyl commented on packages/compiler-cli/src/ngtsc/annotations/src/util.ts:
```diff
@@ -176,3 +176,20 @@ export function forwardRefResolver(
+} {
+  const methods = nodeStaticMethodNames(node, reflector);
+  const ngTemplateGuards = methods.filter(method => method.startsWith('ngTemplateGuard_'))
+                               .map(method => method.split('_', 2)[1]);
```
The mechanism doesn't seem to be generic enough, how could `NgIf` guard injected template, `ngIfThen` and `ngIfElse` simultaneously? Should at least be possible to write:

```typescript
class NgIf {
  static ngTemplateGuard_ngIf<E>(...)
  static ngTemplateGuard_ngIf_ngIfThen<E>(...)
  static ngTemplateGuard_ngIf_ngIfElse<E>(...)
}
```

alxhub replied:
Sure, you can narrow multiple expressions:

```typescript
class NgIf {
  static ngTemplateGuard_ngIf<E>(...)
  static ngTemplateGuard_ngIfThen<E>(...)
  static ngTemplateGuard_ngIfElse<E>(...)
}
```

However in practice this would be useless. Narrowing inside a template scope only happens for the `<ng-template>` that the directive is on. There is no association between `NgIf` and any other templates (even if they're passed in as input values to `ngIfElse`, etc).

Such an association would be difficult to build into the system.

trotyl replied:
> Sure, you can narrow multiple expressions

The problem is not about multiple expressions, but multiple `<ng-template>`:

```html
<!-- items: number[] | null -->
<ng-container *ngIf="items then foo else bar"></ng-container>
<ng-template #foo>{{ items.length }}<!-- No error here --></ng-template>
<ng-template #bar>{{ items.length }}<!-- Throw error here --></ng-template>
```

And to support that there need to be type guard per `@Input()` and per `TemplateRef`, the pattern above I suggested is:

```
static ngTemplateGuard_(propertyNameForExpressionToNarrowWith)_(proprtyNameForTemplateRef)
```

> There is no association between `NgIf` and any other templates

The association were built when assignment happens, so the type-checking can happens there as well.

> Such an association would be difficult to build into the system.

I agree it require some extra work so may not be added in this PR, but it's still quite possible. The basic model is to construct an anonymous type according to actual usage:

```typescript
type anonymous_0 = { items: { length: unknown } }
```

And when the assignment happens check whether the actual type is assignable to usage type, I have made [an informative example here](http://www.typescriptlang.org/play/index.html#src=declare%20function%20nextContext%3CT%3E()%3A%20T%0D%0A%0D%0Atype%20TemplateFn%3CC%2C%20T%3E%20%3D%20(rf%3A%20any%2C%20context%3A%20C%2C%20_%3F%3A%20T)%20%3D%3E%20void%0D%0A%0D%0Atype%20Falsy%20%3D%20false%20%7C%20null%20%7C%20undefined%20%7C%200%20%7C%20''%0D%0A%0D%0Aclass%20NgIf%3CT%3E%20%7B%0D%0A%20%20%20%20ngIf%3A%20T%0D%0A%0D%0A%20%20%20%20static%20ngTemplateGuard_ngIf_ngIfThen%3CE%3E(dir%3A%20any%2C%20expr%3A%20E%20%7C%20Falsy)%3A%20expr%20is%20E%20%7B%0D%0A%20%20%20%20%20%20%20%20return%20true%0D%0A%20%20%20%20%7D%0D%0A%0D%0A%20%20%20%20static%20ngTemplateGuard_ngIf_ngIfElse%3CE%3E(dir%3A%20any%2C%20expr%3A%20E)%3A%20expr%20is%20E%20%26%20Falsy%20%7B%0D%0A%20%20%20%20%20%20%20%20return%20true%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0A%2F*%0D%0A%0D%0A%3Cng-container%20*ngIf%3D%22items%20then%20foo%20else%20bar%22%3E%3C%2Fng-container%3E%0D%0A%3Cng-template%20%23foo%3E%7B%7B%20items.length%20%7D%7D%3C%2Fng-template%3E%0D%0A%3Cng-template%20%23bar%3E%7B%7B%20items.length%20%7D%7D%3C%2Fng-template%3E%0D%0A%0D%0A*%2F%0D%0A%0D%0A%0D%0Atype%20App_Context%20%3D%20%7B%20items%3A%20number%5B%5D%20%7C%20false%20%7D%0D%0A%0D%0Afunction%20App_Template(rf%3A%20any%2C%20context%3A%20App_Context)%20%7B%0D%0A%20%20%20%20const%20expr%20%3D%20context.items%0D%0A%20%20%20%20if%20(NgIf.ngTemplateGuard_ngIf_ngIfThen(null%2C%20expr))%20%7B%0D%0A%20%20%20%20%20%20%20%20foo_template(undefined%2C%20undefined%2C%20%5B%7B%20items%3A%20expr%20%7D%5D)%0D%0A%20%20%20%20%7D%0D%0A%20%20%20%20if%20(NgIf.ngTemplateGuard_ngIf_ngIfElse(null%2C%20expr))%20%7B%0D%0A%20%20%20%20%20%20%20%20bar_template(undefined%2C%20undefined%2C%20%5B%7B%20items%3A%20expr%20%7D%5D)%0D%0A%20%20%20%20%7D%0D%0A%7D%0D%0A%0D%0Atype%20anonymous_0%20%3D%20%7B%20items%3A%20%7B%20length%3A%20unknown%20%7D%20%7D%0D%0A%0D%0Afunction%20foo_template%3CT%20extends%20%5Banonymous_0%5D%3E%0D%0A%20%20%20%20(rf%3A%20any%2C%20context%3A%20void%2C%20_%3F%3A%20T)%0D%0A%7B%0D%0A%20%20%20%20const%20app%20%3D%20nextContext%3CT%5B0%5D%3E()%0D%0A%20%20%20%20const%20usage%20%3D%20app.items.length%0D%0A%7D%0D%0A%0D%0Atype%20anonymous_1%20%3D%20%7B%20items%3A%20%7B%20length%3A%20unknown%20%7D%20%7D%0D%0A%0D%0Afunction%20bar_template%3CT%20extends%20%5Banonymous_1%5D%3E%0D%0A%20%20%20%20(rf%3A%20any%2C%20context%3A%20void%2C%20_%3F%3A%20T)%0D%0A%7B%0D%0A%20%20%20%20const%20app%20%3D%20nextContext%3CT%5B0%5D%3E()%0D%0A%20%20%20%20const%20usage%20%3D%20app.items.length%0D%0A%7D%0D%0A).

alxhub replied:
You are correct - your example would not be type-checked at all in the current design. This is a non-trivial problem to solve.

Is `anonymous_0` generated from the expressions within the template itself? That's an interesting idea. Off the top of my head it seems very limited and error prone, not to mention exceedingly complicated to generate in the first place.

I like your approach of having additional guard functions for property bindings that happen to take templates. With that approach, even local usage of `ngTemplateOutlet` could be correctly inferred. The challenge is that the location an `<ng-template>` is declared can be very different from where it's passed into an input to a directive. For example:

```html
<div *ngIf="blah">
  <!-- deep inside some other structure, we use ngIf -->
  <ng-container *ngIf="items then foo else bar"></ng-container>
</div>

<!-- somewhere up the tree, the templates are actually declared -->
<ng-template #foo>{{ items.length }}<!-- No error here --></ng-template>
<ng-template #bar>{{ items.length }}<!-- Throw error here --></ng-template>
```

This is totally legal, and tricky to handle - the guards you want around the template TCBs aren't available at the location in the template where they're defined.

What I think we'd have to do is have static methods with two dimensions like you suggest. When evaluating the `*ngIf` directive on the `<ng-container>` above, we'd notice that template references are passed to the directive and that narrowing applies. We would assume at that point that the directive intends to render the template, and _move_ the template's TCB inside the appropriate guard, _as if_ it were written at that location anyway.

I think this is a very doable solution, but it's not going to be a priority for now.

mary-poppins commented:
You can preview e25a598 at https://pr26203-e25a598.ngbuilds.io/.

JoostK commented on packages/compiler-cli/test/ngtsc/env.ts:
```diff
@@ -0,0 +1,125 @@
+  tsconfig(extraOpts: {[key: string]: string | boolean} = {}): void {
+    const opts = JSON.stringify({...extraOpts, 'enableIvy': 'ngtsc'});
+    const tsconfig: string =
+        `{"extends": "./tsconfig-base.json", "angularCompilerOptions": ${opts}}`;
```
Why not `JSON.stringify` this whole object literal?

alxhub replied:
No reason - this was moved code, I didn't modify the semantics.

JoostK reviewed (commented):
This is awesome! 🎉

At some point in the future, might it become possible for component inputs to specified as required? I know you can extend the directive's selector to include the required inputs as attribute selectors, but its ergonomics are far from ideal.

mary-poppins commented:
You can preview ca2d072 at https://pr26203-ca2d072.ngbuilds.io/.

alxhub commented on packages/compiler/src/selector.ts:
```diff
@@ -150,29 +150,29 @@ export class CssSelector {
     }
     for (let i = 0; i < cssSelectors.length; i++) {
-      this._addSelectable(cssSelectors[i], callbackCtxt, listContext);
+      this._addSelectable(cssSelectors[i], callbackCtxt as T, listContext);
```
That's the original API - we can refactor this more in the future. In general, this cast is marking the expectation that `callbackCtxt` fits into `T`. If they're using the optional parameter, then `T` should be `|undefined`.

alxhub commented on packages/compiler/src/render3/view/t2_binder.ts:
```diff
@@ -0,0 +1,526 @@
+    //   - references: Map of #references to their targets.
+    const {directives, bindings, references} =
+        DirectiveBinder.apply(target.template, this.directiveMatcher);
+    // Finally, run the TemplateBinder to
```
Fixed.

alxhub commented on packages/compiler/src/render3/view/t2_binder.ts:
```diff
@@ -0,0 +1,526 @@
+      let dirTarget: DirectiveT|null = null;
+
+      // If the reference expression is empty, then it matches the "primary" directive on the node
+      // (if there is one). Otherwise it matches the
```
Fixed.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -0,0 +1,610 @@
+      dir.ngTemplateGuards.forEach(inputName => {
+        // For each template guard function on the directive, look for a binding to that input.
+        const boundInput = tmpl.inputs.find(i => i.name === inputName);
+        if (boundInput !== undefined) {
```
No, we just exclude the guard from the list. Not having a matching guard function just means no narrowing of that input expression takes place.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -0,0 +1,610 @@
+  }
+
+  // By default the guard is simply `true`.
+  let guard: ts.Expression = ts.createTrue();
```
No particular reason. This way all the blocks get generated with the same structure.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts:
```diff
@@ -0,0 +1,610 @@
+  return ts.createVariableStatement(
+      /* modifiers */
+      [
+        // ts.createModifier(ts.SyntaxKind.ConstKeyword)
```
Removed.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/context.ts:
```diff
@@ -0,0 +1,239 @@
+  /**
+   * Record a template for the given component `node`, with a `SelectorMatcher` for directive
+   * matching.
+   * @param node Record a
```
Fixed.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/context.ts:
```diff
@@ -0,0 +1,239 @@
+   * @param template
+   * @param matcher
+   */
+  addTemplate<N>(
```
Fixed.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/context.ts:
```diff
@@ -0,0 +1,239 @@
+    // Record the type check block operation for the template itself.
+    this.addTypeCheckBlock(node.getSourceFile(), node, {
+      boundTarget,
+      fnName: `${node.name!.text}_TypeCheckBlock`,
```
Good point. Fixed.

alxhub commented on packages/compiler-cli/src/ngtsc/typecheck/src/host.ts:
```diff
@@ -0,0 +1,79 @@
+    }
+    if (sf !== undefined) {
+      // Maybe augment the file with type checking code via the `TypeCheckContext`.
+      const augSf = this.context.transform(sf);
```
Very good catch - fixed.

JoostK commented on packages/compiler-cli/src/ngtsc/typecheck/src/host.ts:
```diff
@@ -0,0 +1,91 @@
+      this.augmentedSourceFiles.add(augSf);
+      return augSf;
+    } else {
+      return undefined;
```
@alxhub whoops, this is very much broken now ;)

Instead of a set, I would insert a cache entry such as `{ transformed: false, sourceFile: sourceFile }` into the `sfCache` instead. Seems more efficient, I'd expect.

alxhub replied:
D'oh, of course it is. That's what I get for going too fast.

Fixed for real this time.

mary-poppins commented:
You can preview 57c5c2d at https://pr26203-57c5c2d.ngbuilds.io/.

alxhub committed:
feat(ivy): support template type check narrowing for *ngFor and *ngIf

This commit adds an ngTemplateGuard_ngIf static method to the NgIf
directive and an ngTemplateContextGuard static method to NgFor. The
function of these two static methods is to enable type narrowing
within generated type checking code for consumers of the directives.

mary-poppins commented:
You can preview 3b36369 at https://pr26203-3b36369.ngbuilds.io/.

mary-poppins commented:
You can preview 15f851c at https://pr26203-15f851c.ngbuilds.io/.

mary-poppins commented:
You can preview e28e2e4 at https://pr26203-e28e2e4.ngbuilds.io/.

mary-poppins commented:
You can preview 73f78ab at https://pr26203-73f78ab.ngbuilds.io/.
